### PR TITLE
transformations: Implement stencil unrolling

### PR DESCRIPTION
**Pull request by PapyChacal**
Implement `stencil-unroll`, unrolling a stencil by repeating and offsetting computation in its body and marking it for unrolled lowering.

### CODE REVIEW

File: xdsl/transforms/stencil_unroll.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        return_op = op.region.block.last_op
+        # ReturnOp is ApplyOp's terminator
+        assert isinstance(return_op, ReturnOp)
+
+        # Don't unroll already unrolled stencils.
+        if return_op.unroll is not None:
+            return
+
+        # Don't work on degenerate apply with no result
+        if len(op.results) == 0:
```
Comment (superlopuh) [inline xdsl/transforms/stencil_unroll.py:70]:
```suggestion
        if not op.results:
```

File: xdsl/transforms/stencil_unroll.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+        # If unroll factors list is shorter than the dim, fill with ones from the front
+        unroll = self.unroll_factor.copy()
+        for _ in range(len(unroll), dim):
+            unroll.insert(0, 1)
+
+        # If unroll factors list is longer than the dim, pop from the front to keep
+        # similar semantics
+        for _ in range(dim, len(unroll)):
+            unroll.pop(0)
```
Comment (superlopuh) [inline xdsl/transforms/stencil_unroll.py:86]: Popping or inserting from the front feels off to me, this seems equivalent and a little less spidey-sense-triggering:
```suggestion
        unroll = self.unroll_factor
        if len(unroll) < dim:
            # If unroll factors list is shorter than the dim, fill with ones from the front
            unroll = [1] * (dim - len(unroll)) + unroll
        elif len(unroll) > dim:
            # If unroll factors list is longer than the dim, pop from the front to keep
            # similar semantics
            unroll = unroll[-dim:]
```

Comment (PapyChacal) [reply xdsl/transforms/stencil_unroll.py]: Sure

File: docs/Toy/Toy_Ch0.ipynb
Comment (AntonLydike) [inline docs/Toy/Toy_Ch0.ipynb:1]: Hmm, we seem to have weird formatting changes in these notebooks. Any idea why?

**Review by AntonLydike — APPROVED** [review]
:yum:

**Pull request merged**